src/librustdoc/html/render.rs

```diff
@@ -3476,12 +3476,41 @@ impl<'a> fmt::Display for Sidebar<'a> {
     }
 }
 
+fn get_methods(i: &clean::Impl) -> Vec<String> {
+    i.items.iter().filter_map(|item| {
+        // Maybe check with clean::Visibility::Public as well?
+        if item.visibility.is_some() {
+            match item.inner {
+                clean::MethodItem(_) => {
+                    if let Some(ref name) = item.name {
+                        if name.len() > 0 {
+                            Some(format!("<a href=\"#method.{name}\">{name}</a>", name = name))
+                        } else {
+                            None
+                        }
+                    } else {
+                        None
+                    }
+                }
+                _ => None,
+            }
+        } else {
+            None
+        }
+    }).collect::<Vec<_>>()
+}
+
 fn sidebar_assoc_items(it: &clean::Item) -> String {
     let mut out = String::new();
     let c = cache();
     if let Some(v) = c.impls.get(&it.def_id) {
-        if v.iter().any(|i| i.inner_impl().trait_.is_none()) {
-            out.push_str("<li><a href=\"#methods\">Methods</a></li>");
+        let ret = v.iter()
+            .filter(|i| i.inner_impl().trait_.is_none())
+            .filter_map(|i| Some(get_methods(i.inner_impl()).join("")))
+            .collect::<Vec<_>>().join("");
```

Comment:
1. Isn't `.filter_map(|i| Some(...))` equivalent to `.map(|i| ...)`?
2. You could `.collect::<String>()` directly, instead of `.collect::<Vec<_>>().join("")`

So I guess what you want is 

```rust
v.iter()
    .filter(|i| i.inner_impl().trait_.is_none())
    .flat_map(|i| get_methods(i.inner_impl()))
    .collect::<String>();
```